docs/proposals/PROPOSAL_B.md

```diff
@@ -0,0 +1,287 @@
+# Proposal B - Extend Platform Variants and Introduce Compatibility Artifact
+
+This proposal recommends changes to image spec for more granular image selection and introduces a new image compatibility artifact type.
+Based on the artifact, cluster administrators can verify hosts or prepare the appropriate infrastructure for containers.
+
+## Modifications
+
+### Image Spec
+
+Extend the platform variants with new ARM and AMD64 micro-architecture levels.
+
+| ISA/ABI          | architecture | variant |
+|------------------|--------------|---------|
+| ...              | ...          | ...     |
+| ARM 64-bit, v9   | arm          | v9      |
+| AMD64 64-bit, v2 | amd64        | v2      |
+| AMD64 64-bit, v3 | amd64        | v3      |
+| AMD64 64-bit, v4 | amd64        | v4      |
+
+## Image Compatibility Artifact
+
+!["Image Compatibility Artifact"](img/B_artifact.png "Image Compatibility Artifact")
+
+### Definitions
+
+* image-compatibility artifact - an OCI artifact that stores the compatibility spec.
+* Compatibility Spec - a json object that represents a graph of namespaces.
+* Namespace - a validation area under which users can describe plugin specs that are consumed by plugins (executable binaries) as input.
+* Plugin - an executable binary that extracts data and checks compatibility on a host.
+
+A new imageCompatibility artifact type that contains a layer with a JSON compatibility object.
+The object represents a graph of namespaces.
+Namespace describes the validation area and is another object containing data that provides plugins specification for the corresponding plugins.
+A plugin is an executable binary that validates the host and returns information about success or failure.
+
+### Compatibility Spec
+
+Compatibility spec is a graph that can have multiple roots with different paths.
```

Comment:
I think the way I've been thinking about this (in the context of a graph) is that the compatibility spec exists to provide the base units of information (that are needed to assess compatibility) but there is no requirement of all namespaces passing to go through. This wouldn't work for many cases that only care about one slice of that space (e.g., one specific namespace). The plugin logic is what should determine the compatibility, and use the metadata there. That way if there is a bug that needs to be fixed, you fix the plugin code and don't need to update a gazillion compatibility objects.

Reply:
I'll try to this weekend open up a third proposal that is similar here to create an artifact, but has a slightly different design for how this works. I think I already wrote out some of the idea in my original document from last year so I'll try to formalize that.

Reply:
> This wouldn't work for many cases that only care about one slice of that space (e.g., one specific namespace). The plugin logic is what should determine the compatibility, and use the metadata there.

My angle on that is, once you build a compatibility graph, it should not be common to execute just part of it as together all nodes in the graph provides information if the host is valid or not. To have a full validation check you should execute the whole graph. However, if you decide to execute just a slice then we may add a specific option to the tool to cut the slice of graph and execute.

> The plugin logic is what should determine the compatibility, and use the metadata there.

IMO it should be a plugin logic and its deps in the graph. I prefer to keep this mindset in the proposal.

> I'll try to this weekend open up a third proposal that is similar here to create an artifact, but has a slightly different design for how this works. I think I already wrote out some of the idea in my original document from last year so I'll try to formalize that.

Cool! We will have more perspectives for the discussion and the final implementation.

Reply:
>That way if there is a bug that needs to be fixed, you fix the plugin code and don't need to update a gazillion compatibility objects

very good point.

for the plugin source, we may need to provide similar logic like image tag latest/head, so that the recent plugin can be used in validation, and provide some fallback path if the latest one doesn't work.

other ideas can also be explored.